src/adsbinarynode.jl

```diff
@@ -0,0 +1,402 @@
+    x::SVector{M, TaylorN{T}}, p::SVector{M, Taylor1{TaylorN{T}}}
+    ) where {N, M, T <: Real}
+    isnothing(parent.right) || error("Right child is already assigned")
+    node = ADSBinaryNode{N, M, T}(s, t, x, p, parent.depth + 1, parent)
+    parent.right = node
+end
+
+# AbstractTrees interface
+function children(node::ADSBinaryNode{N, M, T}) where {N, M, T <: Real}
+    if isnothing(node.left) && isnothing(node.right)
+        ()
+    elseif isnothing(node.left) && !isnothing(node.right)
+        (node.right,)
+    elseif !isnothing(node.left) && isnothing(node.right)
+        (node.left,)
+    else
+        (node.left, node.right)
+    end
+end
+
+printnode(io::IO, n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real} = print(io, "t: ", n.t)
+
+nodevalue(n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real} = (n.s, n.t, n.x, n.p)
+
+ParentLinks(::Type{<:ADSBinaryNode{N, M, T}}) where {N, M, T <: Real} = StoredParents()
+
+parent(n::ADSBinaryNode{N, M, T}) where {N, M, T <: Real} = n.parent
+
+NodeType(::Type{<:ADSBinaryNode{N, M, T}}) where {N, M, T <: Real} = HasNodeType()
+nodetype(::Type{<:ADSBinaryNode{N, M, T}}) where {N, M, T <: Real} = ADSBinaryNode{N, M, T}
+
+# For TreeIterator
+IteratorEltype(::Type{<:TreeIterator{ADSBinaryNode{N, M, T}}}) where {N, M, T <: Real} = HasEltype()
+eltype(::Type{<:TreeIterator{ADSBinaryNode{N, M, T}}}) where {N, M, T <: Real} = ADSBinaryNode{N, M, T}
+
+"""
+    countnodes(n::ADSBinaryNode{N, M, T}, k::Int) where {N, M, T <: Real}
+
+Return the number of nodes at depth `k` starting from root node `n`.
+"""
+function countnodes(n::ADSBinaryNode{N, M, T}, k::Int) where {N, M, T <: Real}
+    @assert k >= 0 "k must be greater or equal to 0"
+    if iszero(k)
+        return 1
+    else
+        return countnodes(n.left, k-1) + countnodes(n.right, k-1)
+    end
+end
+
+countnodes(n::Nothing, k::Int) = 0
+
+"""
+    countnodes(n::ADSBinaryNode{N, M, T}, t::T) where {N, M, T <: Real}
+
+Return the number of nodes at time `t` starting from root node `n`.
+"""
+function countnodes(n::ADSBinaryNode{N, M, T}, t::T) where {N, M, T <: Real}
+    if !isnothing(n.left) && (n.t < n.left.t && t < n.t) || (n.t > n.left.t && t > n.t)
+        # t is outside time range
+        return 0
+    else
+        # Count nodes at time t (recursively)
+        return _countnodes(n, t)
+    end
+end
+
+function _countnodes(n::ADSBinaryNode{N, M, T}, t::T) where {N, M, T <: Real}
+    # Count nodes at time t (recursively)
+    if !isnothing(n.parent) && (abs(n.parent.t) <= abs(t) < abs(n.t)) || (t == n.t && isnothing(n.left))
+        return 1
+    else
+        return _countnodes(n.left, t) + _countnodes(n.right, t)
+    end
+end
+
+_countnodes(n::Nothing, t::T) where {T <: Real} = 0
+
+"""
+    timeshift!(n::ADSBinaryNode{N, M, T}, dt::T) where {N, M, T <: Real}
+
+Shift the time of every node in `n` by `dt`.
+"""
+function timeshift!(n::ADSBinaryNode{N, M, T}, dt::T) where {N, M, T <: Real}
```

Comment:
I could not find tests covering `timeshift!` or else being used throught the code, so I'd suggest to either remove this function, or it's deemed useful, to add tests for it.

Reply:
`timeshift!` will be useful in the future to set the epoch of an ADS integration in NEOs. Meanwhile, I've added a couple of tests for this function.